services/shhext/mailservers/connmonitor.go

```diff
@@ -0,0 +1,85 @@
+package mailservers
+
+import (
+	"sync"
+	"time"
+
+	"github.com/ethereum/go-ethereum/log"
+	"github.com/ethereum/go-ethereum/p2p/enode"
+	"github.com/status-im/whisper/whisperv6"
+)
+
+// NewLastUsedConnectionMonitor returns pointer to the instance of LastUsedConnectionMonitor.
+func NewLastUsedConnectionMonitor(ps *PeerStore, cache *Cache, whisper EnvelopeEventSubscbriber) *LastUsedConnectionMonitor {
+	return &LastUsedConnectionMonitor{
+		ps:      ps,
+		cache:   cache,
+		whisper: whisper,
+	}
+}
+
+// LastUsedConnectionMonitor watches relevant events and reflects it in cache.
+type LastUsedConnectionMonitor struct {
+	ps    *PeerStore
+	cache *Cache
+
+	whisper EnvelopeEventSubscbriber
+
+	quit chan struct{}
+	wg   sync.WaitGroup
+}
+
+// Start spins a separate goroutine to watch connections.
+func (mon *LastUsedConnectionMonitor) Start() {
+	mon.quit = make(chan struct{})
+	mon.wg.Add(1)
+	go func() {
+		events := make(chan whisperv6.EnvelopeEvent, whisperEventsBuffer)
+		sub := mon.whisper.SubscribeEnvelopeEvents(events)
+		for {
+			select {
+			case <-mon.quit:
+				sub.Unsubscribe()
+				mon.wg.Done()
+				return
+			case err := <-sub.Err():
```

Comment:
We can potentially have a while loop here.

Reply:
what do you mean?

Reply:
`sub` returns an error, it's handled with `sub = mon.whisper.SubscribeEnvelopeEvents(events)` and the new `sub` returns error again and so on. I am not sure if it's possible but looking at this fragment, it looks like it is.

Reply:
i guess i will exit goroutine. the thing is that this error shouldn't be raised at all. but it exists in the interface